sdk-extensions/logging/src/main/java/io/opentelemetry/sdk/logging/data/LogRecord.java

```diff
@@ -22,19 +24,35 @@ public static LogRecordBuilder builder() {
   }
 
   static LogRecord create(
+      Resource resource,
+      InstrumentationLibraryInfo instrumentationLibraryInfo,
       long timeUnixNano,
       String traceId,
       String spanId,
       int flags,
       Severity severity,
       String severityText,
       String name,
-      AnyValue body,
+      String body,
```

Comment:
Why this deviation?  https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body

Reply:
My use case is not tied to traditional java logging. I would foresee that we would log data, most likely serialized to json. A string therefore is fine, maybe a byte array would do too.

Reply:
Most java logging library have structured logging, it's just not used much.   Additionally, ALL of them have exceptions that should be dealt with.

- [Slf4j LoggingEvent](http://www.slf4j.org/api/org/slf4j/event/LoggingEvent.html) supports key-value pairs.  It's not clear to me these belong as attributes, and if you look at recommended usage, I believe these are likely JSON-like body attributes in practice.
- [Logstash](https://www.innoq.com/en/blog/structured-logging/) added hacky ways to encode key-value pairs that show up in the message and not in the message.
- [Java Logging has parameters](https://docs.oracle.com/javase/7/docs/api/java/util/logging/LogRecord.html#getParameters()), which it's unclear/unspecified if these belong in `attributes` or `body` of log
- [log4j has StringMapMessage](https://logging.apache.org/log4j/log4j-2.9.1/log4j-api/apidocs/index.html?org/apache/logging/log4j/message/StringMapMessage.html) which is used to encode key-value pairs

Those are just the ones I know of when I was reworking the build logging framework for sbt.

I agree with the notion of keeping simplicity and as long as we're ok adding AnyValue back in when we integrate these log messages, I'm on board.

Reply:
I *think* that all the java logging frameworks end up reducing down to String by the time you're implementing a log appender. Yes, the APIs let you send in all sorts of crazy stuff, but you have to provide formatters for that stuff, and the formatters (again, as far as I'm aware) all format to a String by the time we'd be picking up the data to convert to OTLP.

Reply:
One thing that is missing from this whole discussion is specification from the logging sig about how all of these various and sundry logging frameworks' data is supposed to be mapped into OTLP.  Yes, we have inter-format conversion recipes, but that won't help appender implementors for a specific logging framework.  In particular, it's telling that there is no mention of an Exception type in the current spec (although there is a mention of stack traces in the ECS transformation, and that just goes into an attribute).

I fear that if we put `AnyValue` as the body for our LogRecord, it will make it all the more confusing for an appender implementor to know what they're supposed to put where. At least if the body is a `String` it's pretty clear that you want the formatted log message to go there.

All that being said, if there is a compelling use-case for having an arbitrary `AnyValue` thing as the body, then we can add an API for that. To start with, though, I'd like to make the common thing (String) easy, and target making the less common thing possible when/if we find a use-case for it.

Reply:
FWIW, +1 to keeping body as string at least for now. I can't find a difference in meaning between a body with key/value pairs and also having attributes with key/value pairs. Indexing behavior seems the same, index both in general, for example, `StringMapMessage` seems to be the most explicitly a body of key/value pairs at least based on its name, but usages seem to look like attributes, or at least they do need to be indexed the same as attributes would be.

https://github.com/search?q=stringmapmessage&type=code

I have a feeling the idiomatic way of handling this would be to convert all fields of a map to attributes, except for a field named `message` (just by convention, there is no strict meaning to this in the frameworks) which would be copied into body. I've only used JSON logging with Stackdriver before, but I recall it doing something similar to this when scraping.

This is also only assuming it's even possible to do this, though @jkwatson points out appenders may not even get anything other than a string after formatting. I haven't dug into that but that would make it even more compelling to stick to string for now.